theano/gof/utils.py

```diff
@@ -210,6 +211,47 @@ def __str__(self):
 
         return type.__new__(cls, name, bases, dct)
 
+    _cache = {}
+
+    def __call__(cls, *args, **kwargs):
+        _set_cache = cls.__dict__.get('_set_cache', None)
+        if _set_cache:
+            key = [cls]
+            for arg in args:
+                arg = make_hashable(arg)
+                key.append(arg)
+            for k, v in sorted(kwargs.items()):
+                key.append(k)
+                v = make_hashable(v)
```

Comment:
My very first implementation was only caching the ops with valid keys. The benchmarking does not show any significant change in the timing so at least for now it is not causing any problems.

Suppose a simple case that user passes a list instead of a tuple. It's very easy to convert it to a hashable key. Maybe we can remove the ```frozendict ``` or ```hash_from_ndarray``` if you think they might be bottle necks.